### ISSUE TITLE
RunAsync not returning when supplied CancellationToken cancelled

### ISSUE THREAD

**Issue by kierenj** [aspnet/AspNetCore#8501]
Hi, I have some more information around this issue:

https://github.com/aspnet/Hosting/issues/1474

I was using `IWebHost.RunAsync(stop)` and cancelling in a `Console.CancelKeyPress` handler.  Intermittently, this would fail: it would hang rather than return.

However I found another issue - I wasn't setting the `Cancel` flag in the `CancelKeyPress` handler, so the default behaviour is that the app would be exited.  Strangely I would have thought this would have meant less hangs, but running in VS at least, the program wouldn't exit, and it wouldn't return from `RunAsync`.  This was incorrect, I wanted to set `Cancel`, so the process doesn't automatically exit, but I can control it.

Now I'm setting `Cancel` in that handler, things close down naturally every time.  `RunAsync` returns just fine.

**Comment by Tratcher** [comment]
So everything is working as expected now?

**Comment by kierenj** [comment]
I think so, but I wasn’t sure if its as expected - when using Ctrl-C to signal a cancel, but not cancelling in the event handler via setting the Cancel prop, the process hangs.

**Comment by Tratcher** [comment]
Something isn't getting unblocked and breaking into the debugger should show what that is. However, since we know how to do it correctly it's not worth chasing.

Also, why are you implementing CancelKeyPress yourself? `RunAsync()` already does that.

**Comment by anurse** [comment]
Closing this as we haven't heard from you. Please feel free to comment if you're able to get the information we're looking for and we can reopen the issue if we're able to identify the problem.
